### ISSUE TITLE
Add Random instances for tuples

### ISSUE THREAD

**Issue by ndmitchell** [haskell/random#26]
Looking at the interface, I see no reason tuples can't be instances of the Random class, which would make it more useful.

**Comment by cartazio** [comment]
That sounds super reasonable. I'll talk about it with Dominic when we next
catchup.

I assume we probably don't need more Than size 6 or 10 tuples? :)

**Comment by ndmitchell** [comment]
My particular use case was for 4 tuples - specifically, I wanted to call [`seedTFGen`](https://hackage.haskell.org/package/tf-random-0.5/docs/System-Random-TF-Gen.html#v:seedTFGen) with a randomly chosen seed. Agreed, 6 seems sufficient.

**Comment by cartazio** [comment]
@ndmitchell this is a good feature, i'll make it happen, see what i can do...

how would you specify/what to talk about generating samples / intervals?

**Comment by cartazio** [comment]
(i'm amidst revisiting / generalizing the sampler api, though that wont happen for the next release, i presume you want every tuple value, not an "interval/range"?

**Comment by ndmitchell** [comment]
@cartazio I didn't really have any idea about what intervals should mean - but I guess between `(a,b)` and `(c,d)` should probably be equivalent to the first member in the range `a-c` and the second in `b-d`.

**Comment by power-fungus** [comment]
What is the status on this issue? Is there a chance that the pull-request will be merged and an update is published?

**Comment by cartazio** [comment]
Support for tuple sampling will be happening. Stay tuned

**Comment by L-TChen** [comment]
Will this be happening soon? I am really looking forward to this feature. :-)

**Comment by lehins** [comment]
There are now up 7-tuple instances for `Uniform` type class, so by using [`uniform`](https://hackage.haskell.org/package/random-1.2.0/docs/System-Random.html#v:uniform) it is already possible to get random tuples. I don't see a reason why we can't add such instances to `Random` as well. There was too many changes already in `random-1.2`, so we'll add those instances as well as a few other types in the next release.

**Comment by Bodigrim** [comment]
`Random` for tuples requires implementing `randomR` to generate a random value in a range. Should we define `UniformRange` instances for tuples as well then?

**Comment by Shimuuar** [comment]
It's impossible in our current implementation. In order to define such instance we need to be able to count number of elements in range from (a1,b1) to (a2,b2) which in turn require knowing number of values inhabiting `b`. It wouldn't be efficient either number of inhabitants growths *fast*. Consider for example `(Int64,Int64,Int64)`. On top of that instance doesn't make much sense

**Comment by Bodigrim** [comment]
@ndmitchell proposed above that "between (a,b) and (c,d) should probably be equivalent to the first member in the range a-c and the second in b-d". Under this interpretation we can define `UniformRange`, but I'm not quite convinced that this is a right and unambiguous thing to do. (That's basically what I meant to ask in the comment above)

**Comment by lehins** [comment]
I think we can define such instance if we treat `a` and `b` completely independent of each other. For example this could work:
```haskell
instance (UniformRange a, UniformRange b) => UniformRange (a, b) where
  uniformRM ((al, bl), (ah, bh)) g = do
    a <- uniformRM (al, ah) g
    b <- uniformRM (bl, bh) g
    pure (a, b)
```

If we think of the tuples as some related value such complex numbers or something then it definitely doesn't make sense, but above instance (and similar one for `Random`) I believe could be pretty useful. @Shimuuar What's your take on this?

**Comment by Shimuuar** [comment]
That would be easy to implement. And in some sense it _is_ uniform. But it's wrong is we use interpretation: "sample uniformly every value between `a` and `b`". I think it's better to avoid defining weird instances from get-go

**Comment by lehins** [comment]
In that case, I think it would be ok to add such instance for `Random`, but not for `UniformRange`, since former already does not promise uniformity as with types like `Float` and `Integer`.

**Comment by Bodigrim** [comment]
Defining more instances of `Random` than of `UniformRange` will motivate users to use the former and block its eventual deprecation. If we think that this is a handy instance, let's add it to `UniformRange` as well with a proper comment/clarification.

**Comment by lehins** [comment]
I certainly disagree with `Random` being deprecated, even over time! After over two decades of people using it, this would be a wrong move. Moreover, many programming languages have a concept like "`Random`", that produces some standard distribution that does not have a mathematical backing but is useful for programmers. Tuples here is a perfect example.

Users can choose themselves which instance they want to use. One says `Uniform` and it promises the uniform distribution, another one is `Random`, which makes it open to interpretation. Many of the types do coincide and that is fine, IMHO.

**Comment by Bodigrim** [comment]
The problem is that currently `Random` and `randomR` are not open to intepretation: they do claim uniformity. 

https://github.com/haskell/random/blob/11464aa406a3387dc61feab84149feb9f671a5fb/src/System/Random.hs#L189-L190
https://github.com/haskell/random/blob/11464aa406a3387dc61feab84149feb9f671a5fb/src/System/Random.hs#L197-L199

I do not particularly mind lifting this guarantee for `Random`, but I think that in future with a wider adoption of `UniformRange` people would ask for `UniformRange` for tuples as well. I would rather define either both, or none.

**Comment by ndmitchell** [comment]
FWIW, I have no particular desire for a uniform range in 99%+ of cases. I've wanted random tuples a handful of times. I'm perfectly happy with Random.

**Comment by lehins** [comment]
@Bodigrim the whole point of the `Uniform` and `UniformRange` classes coming into existence was for them to be different from `Random` type class. So when "people do ask", I'll be happy to tell them such instance cannot exist, but for now let's not worry about what "people might ask".

I am on the same page with @ndmitchell and @Shimuuar seems to support this idea as well.

So, let's remove the "uniform" promise from `Random` class, because it is already a lie and add some tuple instances for `Random`, but not `UniformRange`.

**Comment by Bodigrim** [comment]
If we communicate the difference between `Random` and `Uniform` clearly and unambiguously, then I'm on board as well.

**Comment by curiousleo** [comment]
@lehins wrote:
> add some tuple instances for `Random`, but not `UniformRange`.

I'm on board with this conclusion.

> So, let's remove the "uniform" promise from `Random` class, because it is already a lie

I don't understand this premise. Which `Random` instances do not produce uniform distributions?

**Comment by lehins** [comment]
> Which Random instances do not produce uniform distributions?

`Integer`, `Float` and `Double`. Calling `random` function for these types will produce values distributed uniformly in a subrange, instead of full range, because as you know full range is infinite

**Comment by lehins** [comment]
There is implementation for this ticket in #72 if anyone feels like giving it a review.

**Comment by lehins** [comment]
Finally this is implemented and is merged into master. These instances will be soon released with `random-1.2.1`
